conans/paths/package_layouts/package_cache_layout.py

```diff
@@ -41,9 +41,19 @@ def __init__(self, base_folder, ref, short_paths, no_lock):
         assert isinstance(ref, ConanFileReference)
         self._ref = ref
         self._base_folder = os.path.normpath(base_folder)
-        self.short_paths = short_paths
+        self._short_paths = short_paths
         self._no_lock = no_lock
 
+    def init_from_metadata(self, loader):
```

Comment:
IMHO, this should be a method of the cache: `layout = cache.init_from_metadata(ref, loader)`, and the layout should be just updated in the context of a `with layout.update_metadata()` call:

```python
layout = cache.init_from_metadata(ref, loader)

with layout.update_metadata() as metadata:
     metadata.recipe.remote = other_remote

assert layout.metadata.recipe.remote == other_remote
```

Reply:
I don't want to need the cache again, I already have a ``layout``, and the layout has everything it needs to recover its data: access to the conanfile path (for short_paths), and access to the metadata (for the revision).

In any case, that is not the biggest issue: I am struggling with some aspects of the modeling of the layout, like the layout containing the ``reference``, it is causing some problems, because in some cases you need it with revisions, and in other cases, having it with revisions is causing errors and misbehavior (in ``search_packages()``, having a revision in ``layout._ref`` with revision filters out packages, and that breaks existing behavior). I will try to fix this first.

Reply:
(Just finishing this conversation before moving to the other one, which is more important, I agree. I'll talk about the `reference` with/out revisions not in this mini-thread)

Everywhere you call the `init_from_metadata`, you have the `cache` object:

This is what you have:
```python
layout = cache.package_layout(ref)
conanfile, _, ref = layout.init_from_metadata(loader)
```

I propose:

```python
layout = cache.init_from_metadata(ref, loader)
```

Of course, I didn't mean to use the `cache` object from inside the `layout` 😅